cockatrice/src/client/ui/layouts/overlap_layout.cpp

```diff
@@ -0,0 +1,486 @@
+{
+
+    // Determine the maximum item dimensions.
+    int maxItemWidth = 0;
+    int maxItemHeight = 0;
+    for (QLayoutItem *item : itemList) {
+        if (item == nullptr) {
+            continue;
+        }
+        if (!item->widget()) {
+            continue;
+        }
+
+        QSize itemSize = item->widget()->sizeHint();
+        maxItemWidth = std::max(maxItemWidth, itemSize.width());
+        maxItemHeight = std::max(maxItemHeight, itemSize.height());
+    }
+
+    // Calculate the overlap offsets.
+    const int extra_for_overlap = (100 - overlapPercentage);
+    const int overlapOffsetWidth =
+        (direction == Qt::Horizontal) ? static_cast<int>(std::round((maxItemWidth / 100.0) * extra_for_overlap)) : 0;
+    const int overlapOffsetHeight =
+        (direction == Qt::Vertical) ? static_cast<int>(std::round((maxItemHeight / 100.0) * extra_for_overlap)) : 0;
+
+    // Variables to hold the total dimensions based on layout orientation.
+    int totalWidth = 0;
+    int totalHeight = 0;
+
+    // Calculate the total size based on the layout direction and constraints.
+    if (direction == Qt::Horizontal) {
+        // Determine the number of columns:
+        // - Use maxColumns if it is greater than 0 (constraint set by the user).
+        // - Otherwise, set the number of columns to the total number of items in the layout.
+        const int numColumns = (maxColumns > 0) ? static_cast<int>(maxColumns) : static_cast<int>(itemList.size());
+
+        // Calculate the extra space required for overlaps between columns:
+        // - Each overlap reduces the effective width of subsequent items.
+        const int extra_space_for_overlaps = overlapOffsetWidth * (numColumns - 1);
+
+        // Total width:
+        // - The first item's width is fully included (maxItemWidth).
+        // - Add the space for all overlaps between adjacent items.
+        totalWidth = maxItemWidth + extra_space_for_overlaps;
+
+        // Determine the number of rows:
+        // - Use maxRows if maxColumns is set (to constrain the number of rows).
+        // - Otherwise, assume a single row.
+        const int numRows =
+            (maxColumns > 0)
+                ? static_cast<int>(ceil(static_cast<double>(itemList.size()) / static_cast<double>(numColumns)))
+                : 1;
+
+        // Total height:
+        // - Multiply the number of rows by the item height (maxItemHeight).
+        // - Subtract the height overlap between rows to avoid double-counting.
+        totalHeight = maxItemHeight * numRows - (numRows - 1) * overlapOffsetHeight;
+    } else if (direction == Qt::Vertical) {
+        // Determine the number of columns:
+        // - Use maxRows to calculate how many columns are needed if a row constraint exists.
+        // - Otherwise, assume a single column.
+        const int numColumns =
+            (maxRows != 0) ? static_cast<int>(ceil(static_cast<double>(itemList.size()) / static_cast<double>(maxRows)))
```

Comment:
I've noticed you use c, stdlib and qt math functions through your pr, I'd be nice if we used one consistently, I'd lean towards using qCeil here.